### DISCUSSION TITLE
Dark theme 🌚

### DISCUSSION THREAD

**Discussion by ghost** [mdn/yari#2539] [Ideas]
https://github.com/mdn/yari/issues/2178

Example:

![Screenshot_20210114-224312](https://user-images.githubusercontent.com/44380480/104652769-ea9c9180-56b9-11eb-936b-dfb8046bfae8.png)

_If the colors can be inverted they have a good contrast ➡️⬅️👁️_

**Comment by peterbe** [comment]
Can you please find the original dark theme issue filed and close this one?

**Comment by peterbe** [comment]
Oh. I see. I'm on the mobile app. This _is_ the discussion. 
I guess we can keep this open and people who want dark mode than react to this discussion.

**Reply by ghost** [reply]
Hey i see discussion in the Android App! 🎉
That's awsome i can comment a lot more 🙂

Yes I vote for dark theme but needs approvment and task and CSS.... 👍

**Reply by Ryuno-Ki** [reply]
Do you have a laptop/desktop device?
You could experiment a dark theme using the [Stylus web extension](https://github.com/openstyles/stylus/#stylus).

A mere invert of the colours won't likely cut it (for typographic reasons).
You might also want to look into https://mozilla.design/ to check for brand guidelines.

**Reply by ghost** [reply]
https://www.google.com/search?q=dark%20theme%20mdn&ie=utf-8&oe=utf-8&client=firefox-b-m

I find a discoruse from 2018 
Anyway what I'm talking Is some PR on minimalist and _voilà 🇫🇷_ 🧞...

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme
(That link talks about dinamic classes but We can Just style `p`, `h1`, etc for starting and the other atoms on demand)

How about `mdn-minimalist-dark` repository....

**Reply by ghost** [reply]
> You might also want to look into https://mozilla.design/ to check for brand guidelines.

That's good.

![Screenshot_20210125-010502.png](https://user-images.githubusercontent.com/44380480/105648015-7cba4c00-5ea9-11eb-8cbe-b5c3e90f6823.png)
![Screenshot_20210125-010440.png](https://user-images.githubusercontent.com/44380480/105648017-7fb53c80-5ea9-11eb-80db-686b676d00a3.png)

**Reply by ghost** [reply]
Why I'm not see a lot of designers commenting in GitHub.
It would be good 👍

**Reply by ghost** [reply]
> Do you have a laptop/desktop device?

Actually not.
I didn't open laptop from about months.
Coding Is impossibile but reading and use GitHub Is ok.
Then is more common to me to use dark theme in mobile because Is a _advertitsed feature_.

**Comment by dannycolin** [comment]
IMO, @mattiapontonio idea to fork mdn-minimalist makes more sense if we want something that can be merged upstream.

**Reply by ghost** [reply]
I agree.
Tagging @schalkneethling 
And then Yari _has to_ (bossy) depends on it.

---

Forking is certainly comfortable. It might also be interesting to start from 0, with 1 CSS file maybe and see what happens ...
https://developer.mozilla.org/en-US/docs/Web/CSS/@import

**Comment by schalkneethling** [comment]
While I agree that a dark theme would be nice to have on MDN Web Docs. We need to consider all the use cases and people this would impact[1]. In the end, it needs to be driven by user choice via `@prefers-color-scheme` but also, we should offer a toggle should you wish to override it.

i.e. based on time of day, on an OS level, the system would have switched to dark mode but, for the moment you wish to disable dark mode.

[1] https://levelup.gitconnected.com/why-dark-mode-causes-more-accessibility-issues-than-it-solves-d2f8359bb46a

**Reply by ghost** [reply]
> we should offer a toggle should you wish to override it.

Yes. Next step.
I see similarities with language selector.

**Reply by Ryuno-Ki** [reply]
Toggle-wise, I'd prefer something like https://adrianroselli.com/ has in his footer.

**Reply by ghost** [reply]
What I'm wondering is if there are browser APIs and then we just need to do the buttons (in Yari) or we need to work in some unstable and questionable system between repositories ...

---

Keep in mind that there would be 2 commitments 

- Manage colors 2 times
- Manage developers who have to use certain classes, make readme, test between different releases ....

**Reply by Ryuno-Ki** [reply]
prefers-color-scheme allows for „light” and „dark” - but you could have a third option „no-preference”, that makes neither decision.

**Reply by ghost** [reply]
> prefers-color-scheme allows for „light” and „dark” - but you could have a third option „no-preference”, that makes neither decision.

And do another repository for this? 🤔 Not sure, 3 times the job Is too much.
However, already two is a struggle

**Reply by dannycolin** [reply]
> And do another repository for this? thinking Not sure, 3 times the job Is too much.

There's no need. We can detect `prefers-color-scheme` with JS. So the functionality to auto-detect and add an option to override the theme could be integrated directly in mdn/yari while the CSS/Sass should be integrated in mdn/mdn-minimalist.

**Reply by ghost** [reply]
Wait now that I think you need 3 repositories: 

1. minimalist Sass repository without colors
2. light theme CSS repository
3. dark theme CSS repository

> with JS

Lets stay on CSS for now...

**Reply by Ryuno-Ki** [reply]
No, it would still be 1. but with two media queries to respect the preferred colour scheme.

**Reply by ghost** [reply]
Yep... 
So we have to move the entire discussion?

**Reply by ghost** [reply]
And open PRs obviously... 😌🙋

**Reply by Ryuno-Ki** [reply]
Like I said: I'd start with creating some mockups (e.g. using the Stylus Web Extension for Firefox or Chrome - on a desktop system at least) and provide them as a source for discussion.

**Reply by ghost** [reply]
https://github.com/mdn/mdn-minimalist/issues/451

I opened a issue.
What i want Is copy and paste from MDN but without custom classes.
I want background black and color White.
And release. 
I Need commits for the scoreboard.

**Reply by ghost** [reply]
Of this PR Is merged We can close this discussion and going forward in minimalist
https://github.com/mdn/mdn-minimalist/pull/452

**Reply by dannycolin** [reply]
@Ryuno-Ki working directly on mdn/mdn-minimalist would be better than Stylus. It's possible to run a local version of MDN for development.

I've done something similar for a few other project (currently even integrating a dark mode in Multi-Account Containers). The solution I used was based on the [.matchMedia()](https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia) method. So I:

1. Get the value of a select element containing three options (auto, light, dark)
2. If auto use `.matchMedia()` to detect the theme
3. if not set the `theme` variable to the `.value` property of the select element
4. Add or change the value of `data-theme` in the body tag with `.dataset()` (See: https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)

In your CSS/Sass, you can have something like:

```lang=css
[data-theme="light"] {
  --page-bg: #fff;
 // etc...
}
[data-theme="dark"] {
  --page-bg: #000;
  // etc...
}
```

**Reply by Ryuno-Ki** [reply]
@dannycolin While I agree, keep in mind that @mattiapontonio does not have a local setup running.
So I try to suggest alternatives for him to start hacking :-)

**Reply by ghost** [reply]
I like it when you clone, install and start everything without any other effort 😌🦧

**Reply by ghost** [reply]
However, I do not believe that there are no Api to request the change of browser settings. It will also be a mess for the language if this is not the case. Can you ask for the location and can't you ask to change the theme? 🤔

**Reply by dannycolin** [reply]
> @dannycolin While I agree, keep in mind that @mattiapontonio does not have a local setup running.
> So I try to suggest alternatives for him to start hacking :-)

But they would have too to upstream the changes.

**Reply by Ryuno-Ki** [reply]
Yes, sure. But you can avoid some back and forth arguments if you can prove, that it's a well-thought out submission.

**Reply by dannycolin** [reply]
IMO, a userstyle isn't a well-thought out submission :)

**Reply by schalkneethling** [reply]
> toggle-wise, I'd prefer something like https://adrianroselli.com/ has in his footer.

That one is pretty neat. Thank you for sharing

**Reply by MJP8** [reply]
I use dark-themed GitHub and Monokai color scheme text editor so I also think **MDN** should have dark theme too.

**Comment by vitaly-zdanevich** [comment]
Maybe at least we can start with dark theme for code examples, instead of writing a lot of code before deployment.

**Comment by ghost** [comment]
MDN burns my retinas, a dark mode toggle would be perfect.

**Comment by sideshowbarker** [comment]
Related request https://github.com/mdn/content/issues/5272

**Comment by skeddles** [comment]
using this website is painful. it's 2021. add dark mode already.

**Comment by rroberrt** [comment]
I wrote a little user script that you can use with an addon script manager like [Tampermonkey](https://addons.mozilla.org/en-US/firefox/addon/tampermonkey/).

This script basically loops over every DOM element, checks its background colour, and if the background colour is brighter than the specified limit, it will apply a formula to it that darkens it.

I guess the biggest problem/concern is&#8212;it doesn't seem to be super fast, so there will be a flash of bright content before it gets dark.

It doesn't look fantastic, but it will definitely darken most of the content. (For some reason it won't affect `.CodeMirror` elements).

Here is the script:

```javascript
// ==UserScript==
// @name         Dark Mode
// @namespace    http://tampermonkey.net/
// @version      0.0.0-alpha.0
// @description  A user script that darkens websites.
// @author       msfninja
// @include      *
// @grant        none
// @run-at       document-start
// ==/UserScript==

// Hide document while page is loading.
document.documentElement.style.backgroundColor = '#000f';
document.documentElement.style.opacity = '0';

// This variable sets the 'limit' of how bright a colour may be
// (a value from 0 to 255) before it's made darker. The sweetspot
// for a decent dark mode seems to be around 20. Setting it to 0
// will make every element black, and setting it to 255 will have
// no effect.
const max = 20;

// Here you can choose what elements should not be affected. You
// can specify classes, IDs or tag names (in the appropriate
// array, presumably).
const exclude = {
	classNames: [
	],
	ids: [
	],
	tagNames: [
		'input'
	]
};

// Starting from here you should probably not touch anything.

// Threshold after which to darken a colour.
const threshold = Math.pow(max, 3);

const
	// querySelectorAll
	qa = selector => {
		return document.querySelectorAll(selector);
	},

	// Checks for class containment.
	contains = element => {

		// Loop over all class names meant to be excluded.
		exclude.classNames.forEach(className => {

			// Does the supplied element contain an excluded class name?
			if (element.classList.contains(className)) {
				// Return `true`.
				return true;
			}
		});

		// Return `false`.
		return false;
	},

	// Applies a darkening formula to a value.
	darken = value => {
		// Return a reduced value that's below the limit (`max`).
		return Math.round(value / 255) * max;
	},

	// Generates an RGBA string.
	generate = array => {

		// Define colours.
		let
			red = darken(array[0]),
			green = darken(array[1]),
			blue = darken(array[2]),

			// If there's an opacity specified, use it. Otherwise
			// use '1' (maximum opacity).
			alpha = array[3] ? array[3] : '1';

		// Return an RGBA string.
		return `rgba(${red}, ${green}, ${blue}, ${alpha})`;
	};

// The thing that does the whole job.
const main = () => {
	// A CSS string of rules
	let css = `.bc-table, .bc-table * { background-color: #000f !important; color: #ffff !important; }`;

	// Loop over all elements in <body>.
	qa('*').forEach(element => {
		// Is the element concerned not one of the excluded ones?
		if (!(
			contains(element) ||
			exclude.ids.includes(element.id) ||
			exclude.tagNames.includes(element.tagName.toLowerCase())
		)) {
			// Set element's colour to white.
			element.style.color = 'rgba(255, 255, 255, 1)';

			// Define some variables.
			let
				// The background of the current element
				background = window.getComputedStyle(element).backgroundColor,

				// RGBA array of the background's RGBA value
				rgba = background.split(/[\(\))]/)[1].split(', '),

				// The average value of the background
				average = rgba[0] * rgba[1] * rgba[2];

			// Is the average above the threshold?
			if (average > threshold) {

				// Generate an ID.
				let id = `id${Math.random().toString().slice(2)}`

				// Fingerprint the element.
				element.classList.add(id);

				// Add CSS rules.
				css += `.${id} { background-color: ${generate(rgba)} !important; }`;
			}
		}
	});

	// Apply CSS rules.
	let style = document.createElement('style');
	if (style.styleSheet) {
		style.styleSheet.cssText = css;
	}
	else {
		style.appendChild(document.createTextNode(css));
	}
	document.head.appendChild(style);
}

// Wait until body is loaded.
document.body.onload = event => {
	// Call `main`.
	main();

	// Unhide the document.
	document.documentElement.style.opacity = '1';
}
```

If someone knows how to optimise this code, I'd like to hear.

**Reply by rroberrt** [reply]
An updated version is available as a gist [here](https://gist.github.com/msfninja/4520fc6bfc97174d0347d635d3143716). I updated the original code here as well. Most issues I initially described seem to be more or less fixed now?..

**Comment by queengooborg** [comment]
I'm +1 for a dark theme for any site or service, and I think that any website, program, etc. should offer a dark mode.  For one of my web development contracts, I led the development of a user-switchable dark theme, defaulting to the value of the `prefers-color-scheme` media query.  I'd be happy to get this one going if this is something we want to go forward with!

(As a current alternative, I have some custom CSS available at https://userstyles.world/style/1268/mozilla-developer-network-dark-theme -- hope you don't mind the shameless self-plug.  :P)

**Comment by sgalanb** [comment]
Hi, I’ve just released an extension that adds a dark mode (with a toggle to choose between on o off, using `localStorage` too) for all MDN Web Docs website and has different VS Code based themes. It isn't perfect but it does the job until an official implementation. It's open-source so feel free to pull-request more themes or any feature/bug.

[Firefox Browser Add-ons](https://addons.mozilla.org/es/firefox/addon/mdn-dark-mode/)

[Chrome Web Apps](https://chrome.google.com/webstore/detail/mdn-dark-mode/jdaepiaaanhlmgoehjnnkhiagfkalhoc)

[Git-Hub repository](https://github.com/sgalanb/MDN-Dark-Mode)

**Comment by queengooborg** [comment]
Just wanted to let everyone know that MDN is planning to incorporate a dark theme into their upcoming redesign, so soon there will be no more need for browser extensions and user CSS!

**Reply by broofa** [reply]
> MDN is planning to incorporate a dark theme into their upcoming redesign

Is there a tracking issue of some sort that we can follow along on?

**Reply by queengooborg** [reply]
They've got a branch that seems to be for the redesign here: https://github.com/mdn/yari/tree/redesign-stage

**Reply by jespertheend** [reply]
Seems the new design has been released. Or at least it has been on my end.

**Reply by lukaszpolowczyk** [reply]
The dots line up badly with a long title:
![obraz](https://user-images.githubusercontent.com/16800535/156199975-e8f7c9ae-a5be-4d16-b0f5-bc779b26047b.png)

It's normal here:
![obraz](https://user-images.githubusercontent.com/16800535/156200067-360560bd-5615-417f-9ad7-a61e3311def6.png)

**Reply by rroberrt** [reply]
Did MDN just completely scrap the ZillaSlab font-face?

**Reply by schalkneethling** [reply]
@msfninja Yeah, we switch too Inter to be inline with the larger Mozilla brand identity. Do you find the new typography hard to read? Is it the typeface itself, the size, or the colours?

**Reply by rroberrt** [reply]
@schalkneethling I don't have problems with reading the typeface, no worries. I just liked it. Tbh, when I first saw the new design, I thought for a second that it looked like a Google product ;)

**Reply by Sashank999** [reply]
Ok this new design is amazing. The font size used for the code can be a little bit bigger.

**Reply by schalkneethling** [reply]
Thank you @Sashank999. We are working on improving the typography, scan ability and general legibility.

**Reply by rroberrt** [reply]
Is there any key combination shortcut to focus on the search bar that you get to see when you click the search icon in the top right corner of the website (see picture below)?

![image](https://user-images.githubusercontent.com/85436086/157908603-f7a3cc1f-4464-420b-a7fd-6a0ed93052a5.png)

<sub>(Or is there a more appropriate place to ask this question?)</sub>

**Reply by dannycolin** [reply]
@msfninja It's `/` but there's currently a bug preventing it to work if the input is hidden. AFAIK, there's already a patch about to or already merged that should be pushed in production with the next release.

**Reply by schalkneethling** [reply]
Yes indeed! We will likely ship a fix this week.
